drivers/gpio/gpio_tca6424a.c

```diff
@@ -0,0 +1,610 @@
+	struct gpio_dt_spec reset_gpio;
+};
+
+/** Runtime driver data */
+struct tca6424a_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+
+	sys_slist_t callbacks;
+	struct k_sem lock;
+	struct k_work work;
+	const struct device *dev;
+	struct gpio_callback int_gpio_cb;
+	struct tca6424a_pins_state pins_state;
+	struct tca6424a_irq_state irq_state;
+};
+
+/**
+ * @brief Gets the state of a specified block of 3 registers from the TCA6424A 
+ * and stores it in specified cache and buffer variables.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param reg Address of the first of 3 registers to be read.
+ * @param cache Pointer to the location to cache the pins states.
+ * @param buf Pointer to the buffer to output the register.
+ *
+ * @retval 0 if successful.
+ * @retval Negative value for error code.
+ */
+static int read_port_regs(const struct device *dev, uint8_t reg,
+			  uint32_t *cache, uint32_t *buf)
+{
+	const struct tca6424a_drv_cfg * const config = dev->config;
+	uint32_t port_data, value;
+	int ret;
+
+	ret = i2c_burst_read_dt(&config->i2c_spec, reg, (uint8_t *)&port_data, 3);
+	if (ret != 0) {
+		LOG_ERR("TCA6424A[0x%X]: error reading register 0x%X (%d)",
+			config->i2c_spec.addr, reg, ret);
+		return ret;
+	}
+
+	value = sys_le24_to_cpu(port_data);
+	*cache = value;
+	*buf = value;
+	LOG_DBG("TCA6424A[0x%X]: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X, "
+		"REG[0x%X] = 0x%X", config->i2c_spec.addr, reg, (*buf & 0xFF), 
+		(reg + 1), ((*buf >> 8) & 0xFF), 
+		(reg + 2), ((*buf >> 16) & 0xFF));
+
+	return 0;
+}
+
+/**
+ * @brief writes to a specified block of 3 registers from the TCA6424A 
+ * and stores it in a specified cache variable.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param reg Address of the first of 3 registers to be written.
+ * @param cache Pointer to the location to cache the pins states.
+ * @param value The pin value to be written into the registers.
+ *
+ * @retval 0 if successful.
+ * @retval Negative value for error code.
+ */
+static int write_port_regs(const struct device *dev, uint8_t reg,
+			   uint32_t *cache, uint32_t value)
+{
+	const struct tca6424a_drv_cfg * const config = dev->config;
+	uint8_t buf[4];
+	int ret;
+	LOG_DBG("TCA6424A[0x%X]: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X, "
+		"REG[0x%X] = 0x%X", config->i2c_spec.addr, reg, (value & 0xFF), 
+		(reg + 1), ((value >> 8) & 0xFF), 
+		(reg + 2), ((value >> 16) & 0xFF));
+
+	buf[0] = reg;
+	sys_put_le24(value, &buf[1]);
+	ret = i2c_write_dt(&config->i2c_spec, buf, sizeof(buf));
+	if (ret == 0) {
+		*cache = value;
+	} else {
+		LOG_ERR("TCA6424A[0x%X]: error writing to register 0x%X "
+			"(%d)", config->i2c_spec.addr, reg, ret);
+	}
+	return ret;
+}
+
+static inline int update_input_regs(const struct device *dev, uint32_t *buf)
+{
+	struct tca6424a_drv_data * const drv_data =
+		(struct tca6424a_drv_data * const)dev->data;
```

Comment:
Is this an unnecessary cast? Same question below.

Reply:
Unnecessary casts have been removed here and everywhere else in the code.